pkg/framework/capi-machinesets.go

```diff
@@ -0,0 +1,195 @@
+
+	return ms, nil
+}
+
+// WaitForCAPIMachineSetsDeleted polls until the given MachineSets are not found, and
+// there are zero Machines found matching the MachineSet's label selector.
+func WaitForCAPIMachineSetsDeleted(ctx context.Context, cl client.Client, machineSets ...*clusterv1.MachineSet) {
+	for _, ms := range machineSets {
+		By(fmt.Sprintf("Waiting for MachineSet %q to be deleted", ms.GetName()))
+		Eventually(func() bool {
+			selector := ms.Spec.Selector
+
+			machines, err := GetCAPIMachines(ctx, cl, &selector)
+			if err != nil || len(machines) != 0 {
+				return false // Still have Machines, or other error.
+			}
+
+			err = cl.Get(ctx, client.ObjectKey{
+				Name:      ms.GetName(),
+				Namespace: ms.GetNamespace(),
+			}, &clusterv1.MachineSet{})
+
+			return apierrors.IsNotFound(err) // MachineSet and Machines were deleted.
+		}, WaitLong, RetryMedium).Should(BeTrue())
+	}
+}
+
+// DeleteCAPIMachineSets deletes the specified machinesets and returns an error on failure.
+func DeleteCAPIMachineSets(ctx context.Context, cl client.Client, machineSets ...*clusterv1.MachineSet) {
+	for _, ms := range machineSets {
+		By(fmt.Sprintf("Deleting MachineSet %q", ms.GetName()))
+		Expect(cl.Delete(ctx, ms)).To(Succeed())
+	}
+}
+
+// WaitForCAPIMachineSet waits for the all Machines belonging to the named
+// MachineSet to enter the "Running" phase, and for all nodes belonging to those
+// Machines to be ready.
+func WaitForCAPIMachineSet(ctx context.Context, cl client.Client, name string) {
+	By(fmt.Sprintf("Waiting for MachineSet machines %q to enter Running phase", name))
+
+	machineSet, err := GetCAPIMachineSet(ctx, cl, name)
+	Expect(err).ToNot(HaveOccurred())
+
+	Eventually(func() error {
+		machines, err := GetCAPIMachinesFromMachineSet(ctx, cl, machineSet)
+		if err != nil {
+			return err
+		}
+
+		replicas := ptr.Deref(machineSet.Spec.Replicas, 0)
+
+		if len(machines) != int(replicas) {
+			return fmt.Errorf("%q: found %d Machines, but MachineSet has %d replicas",
+				name, len(machines), int(replicas))
+		}
+
+		running := FilterCAPIRunningMachines(machines)
+
+		// This could probably be smarter, but seems fine for now.
+		if len(running) != len(machines) {
+			return fmt.Errorf("%q: not all Machines are running: %d of %d",
+				name, len(running), len(machines))
+		}
+
+		for _, m := range running {
+			node, err := GetCAPINodeForMachine(ctx, cl, m)
+			if err != nil {
+				return err
+			}
+
+			if !IsNodeReady(node) {
+				return fmt.Errorf("%s: node is not ready", node.Name)
+			}
+		}
+
+		return nil
+	}, WaitOverLong, RetryMedium).Should(Succeed())
+}
+
+// GetCAPIMachineSet gets a machineset by its name from the default machine API namespace.
+func GetCAPIMachineSet(ctx context.Context, cl client.Client, name string) (*clusterv1.MachineSet, error) {
+	machineSet := &clusterv1.MachineSet{}
+	key := client.ObjectKey{Namespace: ClusterAPINamespace, Name: name}
+
+	Expect(cl.Get(ctx, key, machineSet)).To(Succeed())
```

Comment:
Lets' use Evenutally on all `cl.` (client) actions instead of Expect, so we allow for retry on brief API/network hiccups.